File: backend-app/src/main/java/tom/projects/monkeylog/mapper/ProgramMapper.java
package tom.projects.monkeylog.mapper;

import org.mapstruct.Mapper;
import org.mapstruct.ReportingPolicy;
import tom.projects.monkeylog.dto.program.ProgramResponse;
import tom.projects.monkeylog.model.workout.Program;
import tom.projects.monkeylog.model.workout.Workout;

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses = {WorkoutMapper.class})
public interface ProgramMapper {
    ProgramResponse programToProgramResponse(Program program);
}

File: backend/src/main/kotlin/tom/com/monkeylog/mapper/ProgramMapper.kt
package tom.com.monkeylog.mapper

import tom.com.monkeylog.common.notNull
import tom.com.monkeylog.dto.program.*
import tom.com.monkeylog.model.workout.Program
import tom.com.monkeylog.model.workout.ProgramWeek
import tom.com.monkeylog.model.workout.WorkoutType

fun Program.toResponse() = ProgramResponse(
    id = id.notNull(),
    name = name,
    description = description,
    weeks = this.programWeeks.map(ProgramWeek::toResponse)
)

fun ProgramWeek.toResponse() = ProgramWeekResponse(
    id = id.notNull(),
    name = name,
)

fun ProgramCreateRequest.toEntity() = Program(
    name = name,
    description = description,
    workoutType = WorkoutType.TEMPLATE
)

fun Program.update(programUpdateRequest: ProgramUpdateRequest): Program {
    name = programUpdateRequest.name
    description = programUpdateRequest.description

    return this
}

fun ProgramWeekCreateRequest.toEntity(program: Program) = ProgramWeek(
    name = name,
    program = program
)
